### PR TITLE
ENH - Add configurable distance metric for fuzzy join nearest‑neighbor matching

### LINKED ISSUE #869: ENH - Allow using a different distance for the nearest neighbors in fuzzy join
### Problem Description

ATM we use NearestNeighbors with the l2 distance.
if we could choose the distance to use, then using MinHash as the text encoder and "hamming" as the distance would be an approximation of 1 - Jaccard similarity, which I believed is a common choice for fuzzy joining

### Feature Description

the Joiner would have a "metric" or "distance" parameter that would be forwarded to [NearestNeighbors metric](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors)

### Alternative Solutions

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by sabasiddique1**
REF: #1847 
**Goal**  
Allow users to choose the distance metric used for nearest‑neighbor matching in `fuzzy_join` / `Joiner`, enabling alternatives like cosine or manhattan for different data types.

**Context / Issue**  
Addresses enhancement request to support non‑Euclidean distance metrics for fuzzy join nearest neighbors.

**What changed (step‑by‑step)**  
1. Added a `metric` parameter to `fuzzy_join` and `Joiner` with default `'euclidean'`.  
2. Passed the `metric` through the matching classes to `sklearn.neighbors.NearestNeighbors`.  
3. Documented the new parameter in docstrings with common metric examples.  
4. Added tests covering multiple metrics and invalid metric handling.

**Tests**  
- `pytest skrub/tests/test_fuzzy_join.py -v`

### CODE REVIEW

**Comment by MarieSacksick** [comment]
hello @sabasiddique1!  
Welcome :).  

The CI doesn't pass, could you fix it please? It looks like it's because of the styling, it's explained how to deal with it [here](https://skrub-data.org/stable/CONTRIBUTING.html#formatting-and-pre-commit-checks) in the contribution guide.  
In the mean time, you can put your PR in draft mode, so that it's clear that it's ongoing work. You can do it at the top right of the PR menu:  
image

(I assumed it was still ongoing to I didn't have a look at the full code changes, but I still put a couple of comments on the tests in hope that it will alleviate the number of back and forth to iterate on the PR)

File: skrub/tests/test_fuzzy_join.py [hunk 1]
```diff
@@ -510,3 +510,49 @@ def test_missing_values(df_module):
...
+    # Test with manhattan
+    result_manhattan = fuzzy_join(
+        left, right, on="A", suffix="r", metric='manhattan', add_match_info=False
+    )
+    assert ns.shape(result_manhattan)[0] == 2
+    assert ns.shape(result_manhattan)[1] == 3
+
+    # Test that invalid metric raises error
+    with pytest.raises(ValueError):
+        fuzzy_join(left, right, on="A", metric='invalid_metric')
```
Comment (MarieSacksick) [inline skrub/tests/test_fuzzy_join.py]: This should be in a stand-alone test, to keep tests unitary and dedicated to a specific purpose.

File: skrub/tests/test_fuzzy_join.py [hunk 2]
```diff
@@ -510,3 +510,49 @@ def test_missing_values(df_module):
 
     c = fuzzy_join(b, a, left_on="col3", right_on="col1", add_match_info=True)
     assert ns.shape(c)[0] == len(b)
+def test_fuzzy_join_distance_metrics(df_module):
```
Comment (MarieSacksick) [inline skrub/tests/test_fuzzy_join.py]: instead of repeating the code, you can parameterize it, for instance as it's done [here](https://github.com/skrub-data/skrub/blob/35f4341699f4d725876144450f3fde26d518d6a6/skrub/tests/test_fuzzy_join.py#L14) or [here](https://github.com/skrub-data/skrub/blob/35f4341699f4d725876144450f3fde26d518d6a6/skrub/_reporting/tests/test_utils.py#L68).  
Also, we want to test the values and not only the shape, to make sure that the input "metric" is taken into account, and that the function doesn't always use the default one.

File: skrub/tests/test_fuzzy_join.py [hunk 3]
```diff
@@ -510,3 +510,49 @@ def test_missing_values(df_module):
...
+    """
+    left = df_module.make_dataframe({"A": ["aa", "bb"]})
+    right = df_module.make_dataframe({"A": ["aa", "ba"], "B": [1, 2]})
+
+    # Test with euclidean (default)
+    result_euclidean = fuzzy_join(
+        left, right, on="A", suffix="r", metric='euclidean', add_match_info=False
+    )
+    assert ns.shape(result_euclidean)[0] == 2
+    assert ns.shape(result_euclidean)[1] == 3  # A, Ar, Br
```
Comment (MarieSacksick) [inline skrub/tests/test_fuzzy_join.py]: The comment `# A, Ar, Br` is unclear to me, what do you mean here?

File: skrub/_fuzzy_join.py [hunk 4]
```diff
@@ -36,8 +37,8 @@ def fuzzy_join(
     To identify the best match for each row, values from the matching columns
     (``left_key`` and ``right_key``) are vectorized, i.e. represented by vectors of
     continuous values. Then, the Euclidean distances between these vectors are
```
Comment (rcap107) [inline skrub/_fuzzy_join.py]: The distance here is not necessarily Euclidean anymore, so this needs to be updated

File: skrub/_fuzzy_join.py [hunk 5]
```diff
@@ -120,6 +121,14 @@ def fuzzy_join(
...
         Remove rows for which a match was not found in the right table (i.e. for
         which the nearest neighbor is further than `max_dist`).
+    metric : str, default='euclidean'
+        The distance metric to use for nearest neighbor search.
+        Common options include:
+        - 'euclidean': Standard Euclidean distance (default)
+        - 'manhattan': Sum of absolute differences
+        - 'cosine': Cosine similarity (good for text/TF-IDF vectors)
+        - 'hamming': For binary data
+        See sklearn.neighbors.NearestNeighbors for all available metrics.
```
Comment (rcap107) [inline skrub/_fuzzy_join.py]:
```suggestion
        See :class:`~sklearn.neighbors.NearestNeighbors` for all available metrics.
```

File: skrub/_joiner.py [hunk 6]
```diff
@@ -87,9 +87,23 @@ class Joiner(TransformerMixin, BaseEstimator):
...
+    uses the joined features, or to inspect the result of the join and set
+    a max_dist threshold.
+    metric : str, default='euclidean'
+    The distance metric to use for nearest neighbor search.
+    Common options include:
+    - 'euclidean': Standard Euclidean distance (default)
+    - 'manhattan': Sum of absolute differences
+    - 'cosine': Cosine similarity (good for text/TF-IDF vectors)
+    - 'hamming': For binary data
+    See sklearn.neighbors.NearestNeighbors for all available metrics.
```
Comment (rcap107) [inline skrub/_joiner.py]: Not sure what happened here, but this looks like a copypasting error. The paragraph before `metric` is not relevant here and can be removed, while the section on the metrics should be shortened and refer directly to the sklearn docs for the nearest neighbors (see comment above for the formatting)

File: skrub/_matching.py [hunk 7]
```diff
@@ -15,11 +15,20 @@ class Matching(BaseEstimator):
     nearest match are rescaled. This base class does not apply any rescaling:
     the reference distance is 1.0. Subclasses can override
     ``_get_reference_distances`` to modify the rescaling behavior.
+
+    Parameters
+    ----------
+    metric : str, default='euclidean'
+        The distance metric to use for nearest neighbor search.
+        See sklearn.neighbors.NearestNeighbors for available metrics.
```
Comment (rcap107) [inline skrub/_matching.py]:
```suggestion
        See :class:`~sklearn.neighbors.NearestNeighbors` for all available metrics.  
```

Comment (rcap107) [reply skrub/_matching.py]: It's not a big deal, but none of these classes are public, so there was no need to add the parameters. Good thing to do, however.

**Review by rcap107 — CHANGES_REQUESTED** [review]
Thanks a lot for the PR, @sabasiddique1   
 
I wrote some additional early comments to add to what @MarieSacksick has already mentioned

File: skrub/_joiner.py (hunk 6 above)
Comment (sabasiddique1) [reply skrub/_joiner.py]: Thanks for catching that, I removed the misplaced ‘skrub_Joiner’ paragraph and shortened the metric section to point directly to :class:~sklearn.neighbors.NearestNeighbors. I also updated the fuzzy_join docstring to remove ‘Euclidean distances’ and use the same short link format for consistency.

File: skrub/tests/test_fuzzy_join.py (hunk 3 above)
Comment (sabasiddique1) [reply skrub/tests/test_fuzzy_join.py]: Thanks for pointing that out, I’ll remove or reword the ‘# A, Ar, Br’ comment for clarity.

File: skrub/tests/test_fuzzy_join.py (hunk 2 above)
Comment (sabasiddique1) [reply skrub/tests/test_fuzzy_join.py]: Got it, I’ll parameterize the metric cases and assert the actual matched values (not just shape) so we verify that the metric parameter changes the result.

File: skrub/tests/test_fuzzy_join.py (hunk 1 above)
Comment (sabasiddique1) [reply skrub/tests/test_fuzzy_join.py]: Thanks! I’ll move the invalid‑metric check into its own standalone test to keep it focused.

**Comment by sabasiddique1** [comment]
Can we add the ‘no changelog needed’ label for this PR, or should I add file?

**Comment by rcap107** [comment]
> Can we add the ‘no changelog needed’ label for this PR, or should I add file?

There should be an entry in the "Changes" section of the changelog. All user-facing changes should be reported, and adding a parameter is a user-facing change.

File: skrub/tests/test_fuzzy_join.py (hunk 2 above)
Comment (rcap107) [reply skrub/tests/test_fuzzy_join.py]: This comment was marked as resolved, but the actual problem was not addressed

File: skrub/tests/test_fuzzy_join.py (hunk 3 above)
Comment (rcap107) [reply skrub/tests/test_fuzzy_join.py]: This comment was also resolved without addressing the problem

File: skrub/tests/test_fuzzy_join.py [hunk 8]
```diff
@@ -510,3 +510,96 @@ def test_missing_values(df_module):
...
+
+    # Test with euclidean (default)
+    result_euclidean = fuzzy_join(
+        left, right, on="A", suffix="r", metric="euclidean", add_match_info=False
+    )
+    assert ns.shape(result_euclidean)[0] == 2
+    assert ns.shape(result_euclidean)[1] == 3  # A, Ar, Br
+
+    # Test with cosine
+    result_cosine = fuzzy_join(
```
Comment (rcap107) [inline skrub/tests/test_fuzzy_join.py]: The actual value of the result should be tested against the expected value as a sanity check, for this and all the other metrics.

File: skrub/_joiner.py [hunk 9]
```diff
@@ -87,9 +87,23 @@ class Joiner(TransformerMixin, BaseEstimator):
...
     To identify the best match for each row, values from the matching columns
     (`main_key` and `aux_key`) are vectorized, i.e. represented by vectors of
-    continuous values. Then, the Euclidean distances between these vectors are
-    computed to find, for each main table row, its nearest neighbor within the
-    auxiliary table.
+    continuous values. Then, the distances between these vectors are computed
+    (using the specified metric) computed to find, for each main table row, its
+    nearest neighbor within the auxiliary table.
+
+    Insert some columns whose names start with skrub_Joiner containing
```
Comment (rcap107) [inline skrub/_joiner.py]: This section of the docstring looks like an incorrect copypaste, why is it here?

File: CHANGES.rst [hunk 10]
```diff
@@ -17,6 +17,9 @@ Changes
 - The :class:`StringEncoder` now exposes the ``vocabulary`` parameter from the parent
   :class:`TfidfVectorizer`.
   :pr:`1819` by :user:`Eloi Massoulié <emassoulie>`
+- Added a ``metric`` parameter to :func:`fuzzy_join` and :class:`Joiner` to configure
+  the nearest-neighbor distance used for matching.
```
Comment (rcap107) [inline CHANGES.rst]: Please mention that `metric` can be one of the metrics supported by sklearn.neighbors.NearestNeighbors, with a reference to the its docstring.

**Review by rcap107 — COMMENTED** [review]
Hi @sabasiddique1, there are still a few things to fix. Given the other comments, I thought some of the point had already been addressed. Maybe you forgot to commit your changes?

**Comment by rcap107** [comment]
Hi @sabasiddique1, I've updated the pr manually to avoid trouble with possible merge conflicts. I'm going to review this again today (or tomorrow at the latest). After that we can see if there is still work to do, or if it can be merged.

**Review by rcap107 — APPROVED** [review]
The PR looks good to me, thanks a lot @sabasiddique1! 

I'll leave this open for a few more days in case other maintainers want to take a look, then I'll merge it.

**Comment by rcap107** [comment]
Merging this, thanks @sabasiddique1
